paymentapi/src/main/java/fi/hel/verkkokauppa/payment/paytrail/validation/PaytrailPaymentReturnValidator.java

```diff
@@ -0,0 +1,67 @@
+package fi.hel.verkkokauppa.payment.paytrail.validation;
+
+import fi.hel.verkkokauppa.common.util.StringUtils;
+import fi.hel.verkkokauppa.payment.api.data.PaymentReturnDto;
+import lombok.extern.slf4j.Slf4j;
+import org.helsinki.paytrail.constants.PaymentStatus;
+import org.helsinki.paytrail.service.PaytrailSignatureService;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+import java.util.TreeMap;
+
+@Component
+@Slf4j
+public class PaytrailPaymentReturnValidator {
+
+    @Value("${paytrail.merchant.secret:}")
+    private String secretKey;
+
+
+    public boolean validateChecksum(Map<String,String> checkoutParams, String signature, String orderNumber) {
+        if (buildSignatureFromReturnData(checkoutParams).equals(signature)) {
+            return true;
+        } else {
+            log.debug("paytrail payment validation failed, orderNumber: " + orderNumber);
+            return false;
+        }
+    }
+
+    public PaymentReturnDto validateReturnValues(boolean isValid, String status, String settlementReference) {
+        boolean isPaymentPaid = false;
+        boolean canRetry = false;
+        boolean isAuthorized = false;
+
+        if (isSuccesfull(status) && StringUtils.isNotEmpty(settlementReference)) {
+            isPaymentPaid = true;
+        } else if (isSuccesfull(status) && StringUtils.isEmpty(settlementReference)){
+            isAuthorized = true;
```

Comment:
Should this be false or commented out, because this can change subscription flow